laser_slam/include/laser_slam/benchmarker.hpp

```diff
@@ -2,128 +2,207 @@
 
  private:
-  /// \brief Prevent static class from being instantiated.
-  Benchmarker();
+  // Represents a value and information about the step from which it was collected.
+  struct ValueEntry {
+    ValueEntry(size_t step_id, const TimePoint timestamp, double value)
+      : step_id(step_id), timestamp(timestamp), value(value) {
+    }
+    size_t step_id;
+    TimePoint timestamp;
+    double value;
+  };
 
-  /// \brief Helper class for collecting statistics over a series of measurements.
-  class MeasurementStatistics_ {
+  // Helper class for collecting data and statistics about values.
+  class ValueTopic {
 
    public:
-    /// \brief Add a measurement.
-    void addMeasurement(const double& value);
+    // Adds a value.
+    void addValue(size_t step_id, const TimePoint timestamp, double value);
 
-    /// \brief Compute the mean of the measurements.
+    // Computes the mean of the measurements.
     double getMean() const;
 
-    /// \brief Compute the standard deviation of the measurements.
+    // Computes the standard deviation of the values.
     double getStandardDeviation() const;
 
+    // Gets a reference to the stored data.
+    inline const std::vector<ValueEntry>& getValues() const { return values_; }
+
    private:
-    /// \brief Sum of the measurements, needed for computing mean and standard deviations.
-    double sum_ { 0.0 };
+    // Sum of the values, needed for computing mean and standard deviations.
+    double sum_ = 0.0;
+
+    // Sum of squares of the values, needed for computing the standard deviation.
+    double sum_of_squares_ = 0.0;
 
-    /// \brief Sum of squares of the measurements, needed for computing the standard deviation.
-    double sum_of_squares_ { 0.0 };
+    // The number of values recorded.
+    uint64_t values_count_ = 0;
 
-    /// \brief The number of measurements recorded.
-    uint64_t measurements_count_ { 0 };
+    // The collected values.
+    std::vector<ValueEntry> values_;
   };
 
-  /// \brief Mutex for synchronizing accesses to the benchmarker.
-  static std::mutex mutex_;
+  // Helper functions.
+  static std::string setupAndGetResultsRootDirectory();
+  static TimePoint getFirstValueTimepoint();
+  static double durationToMilliseconds(Duration duration);
+
+  // Mutexes for synchronizing accesses to the benchmarker.
+  static std::mutex value_topics_mutex_;
+  static std::mutex started_measurements_mutex_;
 
-  /// \brief Map containing all the measurements for each named region of code.
-  /// Measurements are sorted in alphabetical order.
-  static std::map<std::string, MeasurementStatistics_> statistics_;
+  // Map containing the values sorted in alphabetical order.
+  static std::map<std::string, ValueTopic> value_topics_;
+
+  // Starting time points of the measurements currently in progress.
+  static std::unordered_map<std::string, TimePoint> started_mesurements_;
+
+  // Timestamp of the step being currently benchmarked.
+  static TimePoint current_timestamp_;
```

Comment:
I am not sure to grasp the need for "steps". I believe that when saving a measurement, the start time and the duration should be all which is required. We can chat about this tomorrow.